p9/messages.go

```diff
@@ -1999,6 +2000,177 @@ func (r *rusymlink) String() string {
+const (
+	ReadLock  LockType = iota
+	WriteLock LockType = 1
+	Unlock    LockType = 2
+)
+
+func (l *LockType) String() string {
+	switch *l {
+	case ReadLock:
+		return "ReadLock"
+	case WriteLock:
+		return "WriteLock"
+	case Unlock:
+		return "Unlock"
+	}
+	return "Unknown lock type"
+}
+
+// A few notes on 9p2000.L and locks.
+// Docs are incomplete and confuse whence and flags. Turns out only flags matter.
+// And whence is not in the packet. And the implementations don't use any of this stuff.
+// They just use flock.
+//
+// Locking over a network almost never works anyway. And note that this packet includes a PID.
+// On a system with, say, 1M cores (these exist) a PID is hardly disambiguating.
+// So, this can't work and is hardly worth doing.
+// "something not worth doing is not worth doing well."
+//
+// This is in some docs,
+// struct flock {
+//     short l_type;  /* Type of lock: F_RDLCK, F_WRLCK, F_UNLCK */
+//     short l_whence;/* How to interpret l_start: SEEK_SET, SEEK_CUR, SEEK_END */
+//     off_t l_start; /* Starting offset for lock */
+//     off_t l_len;   /* Number of bytes to lock */
+//     pid_t l_pid;   /* PID of process blocking our lock (F_GETLK only) */
+// };
+// flags bits are:
+
+// #define P9_LOCK_FLAGS_BLOCK 1    /* blocking request */
+// #define P9_LOCK_FLAGS_RECLAIM 2  /* reserved for future use */
+// client_id is an additional mechanism for uniquely identifying the lock requester and is set to the nodename by the Linux v9fs client.
+
+// LockStat contains lock status result.
+type LockStat uint8
+
+// These are the four current return values for Rlock.
+const (
+	lockOK      LockStat = 0
+	lockBlocked LockStat = 1
+	lockError   LockStat = 2
+	lockGrace   LockStat = 3
+)
+
+// ELockError is the overarching error for Tlock.
+// The 9P2000.L lock errors are basically "something failed",
+// i.e., not conformant with how 9P2000 is supposed to work.
+var ELockError = syscall.Errno(lockError)
+
+// The Linux v9fs client implements the fcntl(F_SETLKW) (blocking)
+// lock request by calling lock with P9_LOCK_FLAGS_BLOCK set. If the
+// response is P9_LOCK_BLOCKED, it retries the lock request in an
+// interruptible loop until status is no longer P9_LOCK_BLOCKED.
+
+// The Linux v9fs client translates BSD advisory locks (flock) to
+// whole-file POSIX record locks. v9fs does not implement mandatory
+// locks and will return ENOLCK if use is attempted.
+
+// Because of POSIX record lock inheritance and upgrade properties,
+// pass-through servers must be implemented carefully.
+
+// tlock is a Tlock message
+type tlock struct {
+	// fid is the fid to lock.
+	fid fid
+
+	Type   LockType // Type of lock: F_RDLCK, F_WRLCK, F_UNLCK */
+	Flags  uint32   // flags, not whence, docs are wrong.
+	Start  uint64   // Starting offset for lock
+	Length uint64   // Number of bytes to lock
+	PID    uint32   // PID of process blocking our lock (F_GETLK only)
+	Client string   // not documented in .h
```

Comment:
"client_id is an additional mechanism for uniquely identifying the lock requester and is set to the nodename by the Linux v9fs client."